### ISSUE TITLE
Change examples for term - occurrenceID

### ISSUE THREAD

**Issue by ahahn-gbif** [tdwg/dwc#491]
## Term documentation change suggestion: move away from triplet combination examples for occurrenceID

* Submitter: ahahn@gbif.org
* Efficacy Justification (why is this change necessary?): In an effort to move towards stable identifiers for occurrence records, we observe that a non-negligible portion of inadvertent ID changes in GBIF-ingested datasets happens because e.g. collection identifier schemes within an institution get consolidated to common pattern across collections. This update then translates into the concatenated occurrenceID values, resulting in an id change. While we are still in a transitional phase where the former triplet of institutionCode, collectionCode, catalogNumber does a meaningful job as a record identifier, we would recommend not to promote this strategy for newly configured datasets, and rather suggest the use of identifiers that do not carry a meaning.
* Demand Justification (if the change is semantic in nature, name at least two organizations that independently need this term): n/a
* Stability Justification (what concerns are there that this might affect existing implementations?): n/a
* Implications for dwciri: namespace (does this change affect a dwciri term version)?: n/a

Current Term definition: https://dwc.tdwg.org/list/#dwc_[term name here]

Proposed attributes of the new term version (Please put actual changes to be implemented in **bold** and ~strikethrough~):

* Term name (in lowerCamelCase for properties, UpperCamelCase for classes): 
* Term label (English, not normative): 
* Organized in Class (e.g., Occurrence, Event, Location, Taxon): 
* Definition of the term (normative): 
* Usage comments (recommendations regarding content, etc., not normative): 

* **Examples (not normative):** 
~http://arctos.database.museum/guid/MSB:Mamm:233627~
000866d2-c177-4648-a200-ead4007051b9
~urn:catalog:UWBM:Bird:89776~

* Refines (identifier of the broader term this term refines; normative): 
* Replaces (identifier of the existing term that would be deprecated and replaced by this term; normative): 
* ABCD 2.06 (XPATH of the equivalent term in ABCD or EFG; not normative):

**Comment by deepreef** [comment]
I STRONGLY support this proposal (as those who know my opinions about this general topic could probably guess).

**Comment by matdillen** [comment]
If we remove these two recommendations, do we state to which terms such identifiers should be mapped instead? I presume this would be materialEntityID or parentMaterialSampleID, but those terms are not accepted yet.

Or is the implication that persistent identifiers with any kind of meaning (including most URIs that are also URLs) should not be used for specimens at all?

**Comment by jdpye** [comment]
Following up on what Abby said, because I was struggling with how to say it and I think she represented it very well, a UUID is not a useful tool for people updating and extending datasets that are actively adding new records and need to distinguish old records from new ones. 

 An example I have is that a listening station detects a coded pinger that is associated to an animal attachment event. Well, the attachment event tells me species, the station deployment tells me place, and the pinger recording event on the instrument tells me time.

 If one of these components changes due to a mistake in recording any piece of that process, when I republish the dataset, I need to locate the entry for the occurrence record I need to correct, using the IDs of the components that I know very well contributed to the creation of the occurrence record. If it's a UUID, I have potentially lost my ability to do that transparently and authoritatively.

 From my experience, which I grant is specific compared to the whole of the database, guidance for this field that would be valuable to a user learning how to create these archives would be to use something that is guaranteed unique, related to the occurrence record's creation, and authoritative to an individual record, and if you don't have something obvious to do that with, use a UUID as a last resort.

**Comment by dbloom** [comment]
I was just responding to this when @peterdesmet's message posted which mirrors my own understanding. This is a request only to have certain examples removed (which I understand, too), not a request to stop the use of certain types of identifiers, such as Triplets.

I will, however, admit to a little confusion about the implications of removing certain types of identifiers from the list of examples. Is there an implication here that GBIF is planning to move away from Triplets and URLs in the occurrenceID field?

Certain collections systems, Arctos being a solid example, are not likely to stop using a URL any time in the foreseeable future. Similarly, as @albenson-usgs described, the Triplet might be the most stable option for many collections with limited staff and technical expertise, especially as we work to include more data from ecologists and biologists not affiliated with traditional museum collections directly. Like Abby, I work with datasets like these, and their publishers, regularly.

I am also curious about the implications here as a GBIF Trainer/Mentor. Historically, Triplets are a recommended option for publishers moving through the BID Programme as well as other GBIF-inspired trainings around the world. Data publishers tend to take the definitions and examples provided in the [DwC Quick Reference Guide](https://dwc.tdwg.org/terms/) quite literally, so I could see some confusion arising among publishers with this change.

**Comment by albenson-usgs** [comment]
I'd really love to discuss this topic in more detail than I think is possible in a GitHub thread (and apologies if I have taken this change request off topic a bit). I'd like to pose it as an unconference topic at the upcoming TDWG. Would folks be interested? Give me a thumbs up emoji if so.

**Comment by dbloom** [comment]
@albenson-usgs That could be a useful conversation. I am unable to attend TDWG this year, so I don't know how participation will work in that unconference setting - we'll have to see how that room is set up.

For purposes here, I would also be curious for a little more clarification from @ahahn-gbif on this matter. No doubt she and the GBIF Data Products folks see many more datasets than we do, but I'm curious to know a little more about the value of removing examples like these from the GBIF perspective. I'm not opposed to the change, I would just like to know more and, like @albenson-usgs, I'm curious what else might be out there that isn't a randomly generated UUID.

**Comment by qgroom** [comment]
Please don't let this thread get sidetracked on the various merits of different identifier systems.
@ahahn-gbif is not recommending a UUID she is removing the implicit recommendation of using a triple.
Triples have long be discredited, they are sometimes not unique, but worse still they totally fail when confronted by the myriad of ways people construct them.  

Therefore I suggest adding an example of a URI as an example of a stable occurence ID

e.g. http://www.botanicalcollections.be/specimen/BR5020224598676V

References about rubbish triples...
[https://doi.org/10.1371/journal.pone.0114069](https://doi.org/10.1371/journal.pone.0114069)
[https://doi.org/10.37044/osf.io/93qf4](https://doi.org/10.37044/osf.io/93qf4)

**Comment by albenson-usgs** [comment]
Ok but the justification for this change states

> In an effort to move towards stable identifiers for occurrence records, we observe that a non-negligible portion of inadvertent ID changes in GBIF-ingested datasets happens because e.g. collection identifier schemes within an institution get consolidated to common pattern across collections. This update then translates into the concatenated occurrenceID values, resulting in an id change.

And what I'm saying is that I don't believe making this change will lead to occurrenceIDs that do not change. If that is the goal then this change, I do not believe, will achieve that result. People will just create a series of letters and numbers instead that will change when they update the dataset.

If what we want to see is stable identifiers for occurrence records then I think we should work on that problem which I don't think will be resolved by only showing UUIDs as the example.

**Comment by dbloom** [comment]
Yes. That is also where my confusion/concern lies. I don't see how the means leads to the desired end.

**Comment by qgroom** [comment]
Except that all the evidence has shown that triples do not work. So far I've not seen any evidence that URIs or DOIs are not stable.

**Comment by MattBlissett** [comment]
> Recommended best practice is to use a persistent, globally unique identifier.

Perhaps expanding the notes would help, to explain why "persistent" and "globally unique" are useful.

> Recommended best practice is to use a persistent, globally unique identifier.
>
> A persistent identifier is one that will not change, allowing others to link to this record in perpetuity.  Avoid including words or numbers that might change, such as the name or abbreviation of a museum department, or a current storage location.
>
> A globally unique identifier works without any other information to identify the occurrence.
>
> A UUID can meet both criteria, so long as the UUID is reliably stored, for example in a collections database.

(I'm not aware of any specific examples Andrea may have in mind.)

**Comment by albenson-usgs** [comment]
Ok here is [an example](https://ipt-obis.gbif.us/resource?r=drytortugasreefvisualcensus2000) where the occurrenceIDs completely changed between V1.0 and V1.1. Admittedly it is me as a naïve data manager just trying to do my best and being told that it must be a UUID. But I am not alone in this- this is what people will do and it will not lead to the result you are wanting.  I think where I struggle with this requirement is how you anticipate it being carried out for small datasets that are managed in Excel. Can you please show me an example of such a dataset that has successfully implemented a DOI or URI scheme for their occurrenceIDs?

**Comment by qgroom** [comment]
Human errors will always happen with whatever system is used. Excel is particular clever at amplifying the chance of human errors.
The problem with triples is that in addition to human errors they are not inherently unique, stable or standardized.

I think we can suggest any of the common systems, including UUIDs, but not triples, because they are not fit for purpose.

**Comment by deepreef** [comment]
Wow... great discussion!  So, I was probably too anemic in my initial response to this proposal.  Although I do happen to support UUIDs as the ideal identifier for our context, that was not the reason I strongly supported this proposal.  My support was for the notion of encouraging opaque identifier -- regardless of whether they are UUIDs or integers or randomly generated character strings or whatever.  When I first started developing biodiversity data management systems (1980s), I was strongly opposed to what we used to call "surrogate primary keys".  Instead, I preferred what we called "natural primary keys" -- that is, identifying some field or combination of fields in the database that (collectively) represented a unique combination of values for each record.  My reasoning was that there was no need to maintain yet another field to uniquely identify each record, when unique identity was self-evident from the actual data-bearing fields themselves.

It didn't take very long for me to realize what a bad mistake that was.  In the decades since then, I've not only embraced "surrogate primary keys" for databases, I've also recognized the importance of NOT using these surrogate primary keys as publicly accessible identifiers.  Instead, they should be optimize for database purposes, such as integrity enforcement and performance for complex joins and stuff like that.  That led me to a database architecture whereby a single data table in my database (called "PK") is responsible for locking in two permanent values:  an integer and a UUID.  The integer serves as the source for internal primary keys for every table in the database, and the UUID is married 1:1 with these integers and is used for representing an identifier for each record whenever the data are exposed externally.

Alas, despite two full-on workshops, multiple mini-confrences and symposia, several robust whitepapers, a couple of publicaitons, and dozens of presentations at TDWG, our community still seems to be stuck when it comes to representing unique identifiers for data records in our exchanged data.  Part of the problem, I think, is that our community got too hung up on the "resolvable" thing.  Basically, to make writing code a little easier, a lot of people wanted to follow the LOD path and conflate the roles of "resolution" (=dereferencing) and "identification" (i.e., unique, stable, persistent).  Some even advocated that every identifier should begin with the characters "http://" (which, of course, results in breaking every single identifier once an SSL certificate is implemented, but I'll leave that one alone for now...)

Coming back to the issue at hand:  there is a strong desire by many to make our identifiers friendly to human eyeballs.  This is one of the reasons everyone likes DOIs more than UUIDs.  But the REAL advantage of DOIs is not that they're easy on the human eyeballs, but because of the robust network of identifier dereferencing that exists (e.g., Crossref, etc.).  It's also one of the reasons why (unfortunately) the outcome of the aforementioned TDWG/GBIF workshops resulted in a recommendation of LSIDs (*sigh*).

Sorry for the rambling context above, but this all leads me to my key point:  We should keep things like triplets and other non-opaque values (i.e. "natural primary keys") in our datasets to make it easier for humans to track things down.  But if we're every going to cross the threshold of data integration and reusability in our community, we really need to get serious about moving towards real identifiers for data records.  And I think this proposal (encouraging opacity for occurrenceID values) is an important step in that direction (though it's certainly not the only step).

Even for people who maintain their data in Excel -- if they can manage a column in their spreadsheets for things like "catalogNumber", why can't they simply add one more column for "occurrenceID", and populate it with some arbitrary and unique and non-information-bearing value (UUID, integer, whatever), then never change that value?  They can use a formula like the one @albenson-usgs gave to initially populate the value, but there's no need to re-run the formula everytime the dataset is exported.

Sorry for the rant -- it is NOT my intention to hijack this discussion to become a general debate about identifiers (which we've had many, many times before).  But I felt the need to respond (indirectly) to some of the comments posted, and expand the explanation for my support for this proposal.

**Comment by Jegelewicz** [comment]
Triples may not be unique, but I don't think you can get more unique than http://arctos.database.museum/guid/MSB:Mamm:233627? A web address must be globally unique or the internet wouldn't function. Just because we include what someone calls a "triple" in it should not make it unworthy of being a GUID. @dustymc

**Comment by deepreef** [comment]
> Triples may not be unique, but I don't think you can get more unique than http://arctos.database.museum/guid/MSB:Mamm:233627? A web address must be globally unique or the internet wouldn't function. Just because we include what someone calls a "triple" in it should not make it unworthy of being a GUID. @dustymc

Sure, you can certainly achive uniqueness; but you break persistence the moment the collection decides to change the collectionCode to "Mam", or someone discovers a typo in the catalogNumber, or if you add an SSL certificate to the website and the identifier changes to "https://arctos.database.museum/guid/MSB:Mamm:233627".

Sure, I know that last one is easily brushed aside because web resolvers usually hande redirection appropriately.  But the point of opacity in identifiers is more about persistence (stability) than it is about uniqueness.

**Comment by albenson-usgs** [comment]
Ok here is [an example](https://ipt-obis.gbif.us/resource?r=noaa_sanctsound_daily_species) of a dataset just freshly published. Code is [here](https://github.com/ioos/bio_data_guide/blob/main/datasets/SanctSound/02_presence_to_occurrence.ipynb). For this one the project was observing sound in the ocean which also includes species observations from those sounds. The project leads have archived all the data and shared it publicly. They are interested to share the species observation portion but the data are not managed with that goal. A data manager separate from the project then took those observations from those sound recordings, aggregated them together (they were spread across multiple datasets), aggregated them so that only one occurrence per day is recorded and aligned them to Darwin Core. An occurrenceID is necessary. The person performing the work is not the holder of the data. I'm truly and honestly curious what those advocating for opaque identifiers would have the data manager do in this instance? Anything that's done will not be stored by the original project.

**Comment by dustymc** [comment]
Yup. ``MSB:Mamm:233627`` isn't a GUID, and if I had my way we'd stop using it even internally. (It works fine internally, but tends to leak and that leads to things like low-quality citations.)

``http://arctos.database.museum/guid/MSB:Mamm:233627`` is a GUID, and is as stable as the Curators care to make it, just like the material is represents. Containing a "triple" doesn't contaminate the identifier. 

We run SSL and it still works.

`` 8088573f-8aba-4f9c-90ab-fefa895b9532 `` is more likely to be unique than ``1`` - unless there's a unique index involved - but it's can't really DO anything and won't generally lead humans to material. (Anyone with the keys to MSB can probably find ``https://arctos.database.museum/guid/MSB:Mamm:233627`` without involving electrons.)

>break persistence the moment the collection decides to change the collectionCode to "Mam", 

That's not a collectionCode, so nope, it'll survive that change no problem. It might devolve to a UUID-like "probably unique but doesn't DO STUFF" identifier if the HTTP protocol goes away or something, but that wouldn't remove its identifier-ness.

DOIs might keep DOING STUFF through events that could remove functionality from URLs, but they're also more work to maintain. There are a few (~20K - example:  https://doi.org/10.7299/X7C829FR ) assigned to records in Arctos, there's no barrier to using them I think the cost/benefit analysis just tends to come up lacking except in some specific situations.

>someone discovers a typo in the catalogNumber,

That can be dealt with in the same sort of way that discovering a problem with the labeling of physical items can be dealt with, and doesn't do anything to the identifier. http://arctos.database.museum/guid/DGR:Mamm:10002772 is an example of a record that's been recataloged (not quite a typo, still a change).

>more about persistence (stability) than it is about uniqueness.

Both are necessary if one wants to DO STUFF at scale.

**Comment by Jegelewicz** [comment]
In the grand scheme of things who cares what the recommendations are? People will do whatever until there is an affordable and manageable path that allows them to do what is necessary.

**Comment by deepreef** [comment]
Sure, as long as "http://arctos.database.museum/guid/MSB:Mamm:233627" is minted once, and remains as such always even if someone realizes the catalog number is actually 233672 for this record -- that's great.  This, of course, requires that the identifier be stored as a literal value (not contstucted automatically from the constituent parts at export time) -- and I wonder how often that's done.  Even if it is done, in my experience there is an overwhelming temptation to want to "correct" the identifier to "http://arctos.database.museum/guid/MSB:Mamm:233672" (when the catalog number error is discovered).  But as long as the data manager can resist that urge, then everything should be OK.

As for conflating the functions of identity with "DOING STUFF", I'll refer to [here](http://bioguid.org/about).  I've written elsewhere (extensively) about why the roles of identity and doing stuff should not be conflated; but that doesn't mean I'm right.

**Comment by tucotuco** [comment]
Thank you all for this thorough and fundamental perspective. The "Darwin Core Triplet" example was included exactly to reduce obstacles to data sharing. It is good to have this additional experience from another community of practice.

**Comment by EstebanMH-SiB** [comment]
Just to add one more example, from [@SiBColombia](https://github.com/SiBColombia) we have used the triplet combination to create stable occurrenceIDs in a successful way for a couple of years, not just for collections but for everything and is our recommendation to publishers [examples in spanish](https://biodiversidad.co/elementos-darwin-core#occurrenceID). 
Sure, sometimes it is impossible to reconcile them and we are forced to change them, but it is less than 5% of the time.

If we tried to change it, virtually no one from our publishers is in the technical capacity to generate stable identities on their own and keep them from version to version, it will just create more difficulties and a new set of identifiers every time, without the possibility of at least trying to match them.

So, we do not endorse this proposal and I am under the impression that there is no consensus and we should not change the examples.

**Comment by ekrimmel** [comment]
> So, we do not endorse this proposal and I am under the impression that there is no consensus and we should not change the examples.

I would second this from @EstebanMH-SiB. If anything, I think _adding_ examples might be helpful. Users in need of a simpler solution could still find such an example, but they'd also be exposed to variations of "more ideal" identifiers. In a way, adding examples may achieve @ahahn-gbif's original intent because it dilutes each of the individual examples.

- ~000866d2-c177-4648-a200-ead4007051b9~ _(current UUID example that we maybe replace with the example directly below)_
- 3af2b96d2-a8a1-47c5-9895-b0af03b21674 _(UUID example)_
- http://n2t.net/ark:/65665/3af2b96d2-a8a1-47c5-9895-b0af03b21674 _(example of the same UUID that is also a resolvable PID)_
- http://arctos.database.museum/guid/MSB:Mamm:233627 _(current example, and a resolvable PID)_
- https://www.botanicalcollections.be/specimen/BR5020224598676V _(another resolvable PID)_
- urn:catalog:UWBM:Bird:89776 _(current example, triplet style)_
- http://doi.org/10.58151/NHB002N5T _(example of an IGSN PID made resolvable as a DOI)_

**Comment by tucotuco** [comment]
I like the idea of added resolvable PIDs, however, we should take the opportunity to provide these examples in the correct context. It looks to me as though all of the examples should actually be materialEntityIDs, not occurrenceIDs. I would add them to materialEntityID examples.

**Comment by ekrimmel** [comment]
@tucotuco I agree with your comment about providing these examples in the correct context, but I think the relationships between identifiers (_organismID_, _occurrenceID_, _materialEntityID_, _materialSampleID_) make doing so require some extra thought. Which you and/or others involved in the GBIF Unified Model work have perhaps already done?

In a model world, could the same identifier ever be used as a value for both _occurrenceID_ and _materialEntityID_? If the primary activity of a collecting event is material gathering that results in a cataloged object, let's assume the object is assiged a catalog number and that it is an identifier which maps conceptually best to _materialEntityID_. An occurrence was never intentionally recorded, but for the purposed of publishing, an occurrence can easily be presumed from existing data. For the sake of thinking this through, let's pretend we have a few options with some real existing data:

1. _occurrenceID_ = http://n2t.net/ark:/65665/3af2b96d2-a8a1-47c5-9895-b0af03b21674, and _materialEntityID_ = http://n2t.net/ark:/65665/3af2b96d2-a8a1-47c5-9895-b0af03b21674.
2. _occurrenceID_ = http://n2t.net/ark:/65665/3af2b96d2-a8a1-47c5-9895-b0af03b21674, and _materialEntityID_ = USNM PAL 652656. The comments for _materialEntityID_ state: "The identifier is **bound to a physical object** (the dwc:MaterialEntity) as opposed to a particular digital record (representation) of that physical object," so perhaps better to use "USNM PAL 652656" because this is what is physically attached to the cataloged object? Using the ARK as an occurrenceID could work, although not if the ARK refers to a cataloged object and the object (in this case a fossil) consists of multiple organisms.
3. _occurrenceID_ = [some as yet uncreated identifier], catalogNumber = USNM PAL 652656, and _materialEntityID_ = http://n2t.net/ark:/65665/3af2b96d2-a8a1-47c5-9895-b0af03b21674. But then, where is this occurrenceID value being created and stored? This isn't a concept most CMS currently handle distinct from either the cataloged object or the collecting event. I think this is the crux of the issue I see, where our current DwC/IPT-centric publishing model encourages data providers to equate occurrence with cataloged object.

Any thoughts about or preferences amongst the options above?

**Comment by ekrimmel** [comment]
Thanks so much for the detailed response above, @tucotuco. I like where your head is at and generally agree that working through some growing pains is worth it for the community to get to a new level semantically.

> I think all examples that are not for dwc:Occurrences should not be included.

This seems like the best way to resolve the issue at hand to me. It undersells the complications and growing pains data publishers will need to resolve for their current dwc:occurrenceID workflows but should help prevent new confusion. Implementing it might look like changing up the examples as follows:

- ~http://arctos.database.museum/guid/MSB:Mamm:233627~ _(current example that refers to a physical specimen, i.e. would more appropriately be an example for dwc:materialEntityID)_
- https://arctos.database.museum/guid/MSB:Mamm:231731?seid=966527 _(resolvable PID, references one dwc:Occurrence associated with a dwc:Organism that also has a second dwc:Occurrence)_
- ~000866d2-c177-4648-a200-ead4007051b9~ _(current example that refers to a physical specimen, i.e. would more appropriately be an example for dwc:materialEntityID)_
- ~urn:catalog:UWBM:Bird:89776~ _(current example that refers to a physical specimen, i.e. would more appropriately be an example for dwc:materialEntityID)_

We might consider also including the following pair, if only to have more than one example:
- 3854b337a-da0b-483f-a1c7-107403bdbb40 _(UUID example)_
- http://n2t.net/ark:/65665/3854b337a-da0b-483f-a1c7-107403bdbb40 _(example of the same UUID that is also a resolvable PID; arguably more appropriate as a dwc:materialEntityID from the institutional process perspective, but purely as an example this could accurately refer to an occurrence because the physical object contains fossils from two organisms which are each given a unique ARK)_

Ironically (only because so much time has been spent discussing), this brings us essentially back to where @ahahn-gbif  wanted to be: without any DWC triplet examples. Ha.
